packages/block-library/src/navigation-link/index.php

```diff
@@ -327,69 +391,42 @@ function build_variation_for_navigation_link( $entity, $kind ) {
+ * Returns the dynamically built variations for all post-types and taxonomies.
  *
- * @param array $variation Variation array from build_variation_for_navigation_link.
- * @return void
- */
-function block_core_navigation_link_register_variation( $variation ) {
-	// Directly set the variations on the registered block type
-	// because there's no server side registration for variations (see #47170).
-	$navigation_block_type = WP_Block_Type_Registry::get_instance()->get_registered( 'core/navigation-link' );
-	// If the block is not registered yet, bail early.
-	// Variation will be registered in register_block_core_navigation_link then.
-	if ( ! $navigation_block_type ) {
-		return;
-	}
-
-	$navigation_block_type->variations = array_merge(
-		$navigation_block_type->variations,
-		array( $variation )
-	);
-}
-
-/**
- * Unregister a variation for a post type / taxonomy for the navigation link block.
+ * @since 6.5.0
  *
- * @param string $name Name of the post type / taxonomy (which was used as variation name).
- * @return void
+ * @param array         $variations Array of registered variations for a block type.
+ * @param WP_Block_Type $block_type The full block type object.
  */
-function block_core_navigation_link_unregister_variation( $name ) {
-	// Directly get the variations from the registered block type
-	// because there's no server side (un)registration for variations (see #47170).
-	$navigation_block_type = WP_Block_Type_Registry::get_instance()->get_registered( 'core/navigation-link' );
-	// If the block is not registered (yet), there's no need to remove a variation.
-	if ( ! $navigation_block_type || empty( $navigation_block_type->variations ) ) {
-		return;
+function block_core_navigation_link_filter_variations( $variations, $block_type ) {
+	if ( 'core/navigation-link' !== $block_type->name ) {
+		return $variations;
 	}
-	$variations = $navigation_block_type->variations;
-	// Search for the variation and remove it from the array.
-	foreach ( $variations as $i => $variation ) {
-		if ( $variation['name'] === $name ) {
-			unset( $variations[ $i ] );
-			break;
-		}
-	}
-	// Reindex array after removing one variation.
-	$navigation_block_type->variations = array_values( $variations );
+
+	$generated_variations = block_core_navigation_link_build_variations();
+	// TODO: Mabe check if there are already variations for this post type/taxonomy?
+	return array_merge( $variations, $generated_variations );
 }
 
 /**
  * Returns an array of variations for the navigation link block.
  *
+ * @since 6.5.0
+ *
  * @return array
  */
 function block_core_navigation_link_build_variations() {
-	// This will only handle post types and taxonomies registered until this point (init on priority 9).
-	// See action hooks below for other post types and taxonomies.
-	// See https://github.com/WordPress/gutenberg/issues/53826 for details.
+	// This will only handle post types and taxonomies registered until this point.
```

Comment:
> until this point

Can we qualify / add more detail to this so it's clearer?

Reply:
I think this comment is not needed @gaambo as now we're handling everything in one go since `get_block_type_variations` handles registration every time we want to list variations right?